### PR TITLE
Bugfix/enable built in element extension

### LINKED ISSUE #607: Extending builtin elements not supported
## Are you reporting a bug or a feature?

 - [x] Bug
 - [ ] Feature

## Expected Behavior

The [`CustomElementRegistry.define()` API](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/define) specifies the [ability to extend builtin elements](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/define#Customized_built-in_element).
So it should be working as specified.

## Actual Behavior

Extending builtin elements isn't fully supported by:
- [not all browsers support it and safari even refuses to implement it](https://developers.google.com/web/fundamentals/web-components/customelements#extendhtml) 😨 
   > **Note:** Only Chrome 67 supports customized built-in elements ([status](https://www.chromestatus.com/feature/4670146924773376)) right now. Edge and Firefox will implement it, but Safari has chosen not to implement it. This is unfortunate for accessibility and progressive enhancement. If you think extending native HTML elements is useful, voice your thoughts on [509](https://github.com/w3c/webcomponents/issues/509) and [662](https://github.com/w3c/webcomponents/issues/662) on Github.
- https://github.com/w3c/webcomponents/issues/727
- the polyfill does not implement it https://github.com/webcomponents/custom-elements/issues/13

## Steps to Reproduce the Problem

Currently we face this issue especially with tables #546 but it's also kinda existing with `<axa-button>`

### PR DESCRIPTION
**Pull request by AndyOGo**
Fixes #607 .
Blocked by #649 #639 https://github.com/choojs/nanohtml/issues/136

Changes proposed in this pull request:

 - replaced the existing polyfill with a much better, battle tested and feature complete polyfill (including built-ins support) from [WebReflection](https://github.com/WebReflection)
 - all kudos go to [Andrea Giammarchi](https://github.com/WebReflection)
- upgrade create amo script
- upgrades `withReact` to support extended builtin elements

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
 
 # How Has This Been Tested?
 
 
 
 # Checklist:
 
 - [ ] My code follows the style guidelines of this project
 - [ ] I have performed a self-review of my own code
 - [ ] I have commented my code, particularly in hard-to-understand areas
 - [ ] I have made corresponding changes to the documentation
 - [ ] My changes generate no new warnings
 - [ ] I have added tests that prove my fix is effective or that my feature works
 - [ ] New and existing unit tests pass locally with my changes
 - [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/app/app.js [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import '@webcomponents/webcomponentsjs/bundles/webcomponents-ce';
+import 'document-register-element/build/document-register-element'; // ES2015
```
Comment (AndyOGo) [inline src/app/app.js]: Well that is broken...
https://github.com/WebReflection/document-register-element/issues/155

File: .babelrc [hunk 2]
```diff
@@ -28,6 +27,88 @@
       "include": [
         "NODE_ENV"
       ]
+    }],
+    ["babel-plugin-transform-builtin-classes", {
```
Comment (AndyOGo) [inline .babelrc]: This awesome plugin is key to make it work
https://github.com/WebReflection/babel-plugin-transform-builtin-classes

**Comment by LucaMele** [comment]
Please Lets make a concrete test in this PR:

- Lets create a extension like `<table is="axa-table"></table>`
- Lets also create a `patterns-library-example` using that table in react.

I would merge it in this PR and when approved, make a update date and communicate it on slack. For example: "Please note, we are closing our first stable release. We realised that we cannot live anymore without supporting Custom Element Builtins. Therefore we will do Breaking change merge friday the XX.YY.KKKK".

**Comment by LucaMele** [comment]
let me know when everything is done and i will review :) also dont forget the table implementation

**Comment by LucaMele** [comment]
I will pay you a hot chocolate or capuccino with lots of milk for this one! Thank u man!

File: .eslintrc [hunk 3]
```diff
@@ -25,11 +25,87 @@
     }]
   },
   "globals": {
+    "HTMLAllCollection": true,
```
Comment (AndyOGo) [inline .eslintrc]: @LucaMele 
We should double check which native constructors we really need.
I would argue that all `HTML*Element` constructors are sufficient.

File: src/js/with-react.js [hunk 4]
```diff
@@ -52,8 +52,8 @@ const withReact = (WebComponent, { pure = true, passive = false } = {}) => {
   // hence it's tagName could only be resolved lazily
   // ref: https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry
   // therefor we don't instantiate it, but rather use a statically defined tagName property
-  const { tagName } = WebComponent;
-  const displayName = `${camelize(tagName)}React`;
+  const { tagName, buildinTagName } = WebComponent;
```
Comment (AndyOGo) [inline src/js/with-react.js]: Needed to support custom builtin elements with react

File: stack/tasks/create-a-m-o.js [hunk 5]
```diff
@@ -33,13 +34,170 @@ console.log(chalk.cyan(outdent`
...
+let isBuiltin = false;
+let buildinElement = '';
 
 const mapElement = {
   a: 'ATOM📗',
   m: 'MOLECULE📘',
   o: 'ORGANISM📙',
 };
 
+const getNativeElementConstructor = (tag) => {
```
Comment (AndyOGo) [inline stack/tasks/create-a-m-o.js]: This is a huge list, maybe we put it into a separate file 💣

File: stack/tasks/create-a-m-o.js [hunk 6]
```diff
@@ -268,29 +467,76 @@ const createBoilerplate = (_name) => {
   }
 };
 
+// @todo: this should be a clean state machine
```
Comment (AndyOGo) [inline stack/tasks/create-a-m-o.js]: This is really key, may it code be made cleaner

**Comment by AndyOGo** [comment]
Yeeeess [`nanohtml v1.3.0`](https://github.com/choojs/nanohtml/releases/tag/v1.3.0) is out and works as expected 🚀

**Comment by LucaMele** [comment]
There are some merge conflicts

**Comment by AndyOGo** [comment]
just rebased and resolved all merge conflicts

**Comment by LucaMele** [comment]
:) @AndyOGo still on here :) i will test it tomorrow :)

**Comment by LucaMele** [comment]
I will review this today afternoon/evening

File: stack/tasks/create-a-m-o.js [hunk 7]
```diff
@@ -33,13 +33,170 @@ console.log(chalk.cyan(outdent`
   `));
 
 let element = '';
+let componentName = '';
+let isBuiltin = false;
+let buildinElement = '';
```
Comment (jackblackCH) [inline stack/tasks/create-a-m-o.js]: `buildinElement` -> `builtInElement`

File: stack/tasks/create-a-m-o.js [hunk 8]
```diff
@@ -33,13 +33,170 @@ console.log(chalk.cyan(outdent`
   `));
 
 let element = '';
+let componentName = '';
+let isBuiltin = false;
```
Comment (jackblackCH) [inline stack/tasks/create-a-m-o.js]: `isBuiltin ` -> `isBuiltIn `

File: stack/tasks/create-a-m-o.js [hunk 9]
```diff
@@ -68,20 +241,32 @@ const getClassName = _name => `AXA${camelCase(_name)}`;
 
 const writeIndexJs = (path, _name) => {
   const className = getClassName(_name);
+  const nativeElementConstructor = getNativeElementConstructor(buildinElement);
```
Comment (jackblackCH) [inline stack/tasks/create-a-m-o.js]: `buildinElement ` -> `builtInElement `

File: stack/tasks/create-a-m-o.js [hunk 10]
```diff
@@ -95,9 +280,12 @@ const writeIndexJs = (path, _name) => {
         // static get observedAttributes() {
         //  return ['classes'];
         // }
-
-        constructor() {
-          super({ styles, template });
+
```
Comment (LucaMele) [inline stack/tasks/create-a-m-o.js]: to much whitespace

File: stack/tasks/create-a-m-o.js [hunk 11]
```diff
@@ -168,10 +360,16 @@ const writeIndexScss = (path, _name) => {
...
+    `;
+
   fs.writeFileSync(
     `${path}/_preview.html`,
-    outdent`<axa-${_name} classes="${element}-${_name}"></axa-${_name}>
-    `
+    markup
     , handleError,
   );
   fs.writeFileSync(
```
Comment (LucaMele) [inline stack/tasks/create-a-m-o.js]: article i9in built ins maybe should be remobved

Comment (LucaMele) [reply stack/tasks/create-a-m-o.js]: and also display block on the scss file

File: stack/tasks/create-a-m-o.js [hunk 12]
```diff
@@ -68,20 +241,32 @@ const getClassName = _name => `AXA${camelCase(_name)}`;
...
+  const nativeElementConstructor = getNativeElementConstructor(buildinElement);
+  const BaseConstructor = isBuiltin ? `${nativeElementConstructor}Base` : 'BaseComponentGlobal';
 
   fs.writeFileSync(
     `${path}/index.js`,
     outdent`import PropTypes from '../../js/prop-types'; // eslint-disable-next-line import/first
 
-      import BaseComponentGlobal from '../../js/abstract/base-component-global';
+      ${isBuiltin ?
+        "import { withBase, withAllHocs } from '../../js/abstract/hocs';"
```
Comment (LucaMele) [inline stack/tasks/create-a-m-o.js]: please use baseglobal as default

File: stack/tasks/create-a-m-o.js (hunk 11 above)
Comment (AndyOGo) [reply stack/tasks/create-a-m-o.js]: Okay, `<article>` only used for non-built-in organisms, else `<div>`???

File: stack/tasks/create-a-m-o.js (hunk 7 above)
Comment (AndyOGo) [reply stack/tasks/create-a-m-o.js]: done

File: CONTRIBUTING.md [hunk 13]
```diff
@@ -214,8 +215,11 @@ A [custom element](https://developer.mozilla.org/en-US/docs/Web/Web_Components/U
 
 #### `constructor()`
 
-The constructor can be used to setup stuff like, establishing contexts, event handlers, observers, defining a shadow root, but never for DOM manipulation.
-It always starts by calling `super()` so that the correct prototype chain is established.
+We never use the ~~constructor~~ directly, ìnstead we call [`init()`](#init) for you with the proper context!
+
+#### `init()`
+The `init()` method can be used to setup stuff like, establishing contexts, event handlers, observers, defining a shadow root, but never for DOM manipulation.
```
[the remark this answers is not in the document]
Comment (AndyOGo) [reply CONTRIBUTING.md]: done

File: stack/tasks/create-a-m-o.js (hunk 8 above)
Comment (AndyOGo) [reply stack/tasks/create-a-m-o.js]: done

File: stack/tasks/create-a-m-o.js (hunk 9 above)
Comment (AndyOGo) [reply stack/tasks/create-a-m-o.js]: done

File: stack/tasks/create-a-m-o.js (hunk 10 above)
Comment (AndyOGo) [reply stack/tasks/create-a-m-o.js]: done

File: stack/tasks/create-a-m-o.js (hunk 12 above)
Comment (AndyOGo) [reply stack/tasks/create-a-m-o.js]: done

**Comment by LucaMele** [comment]
Please communicate the breaking changes and the new API (`constructor` vs `init`)

**Pull request merged**